src/BaseExtractor.php

```diff
@@ -0,0 +1,321 @@
+            $tableManifestOptions
+        );
+    }
+
+    protected function createOutputCsv(string $outputTable): CsvWriter
+    {
+        return new CsvWriter($this->getOutputFilePath($outputTable));
+    }
+
+    protected function createSshTunnel(SshParameters $sshParameters, DatabaseParameters $dbConfig): DatabaseParameters
+    {
+        $privateKey = $sshParameters->getKeys()['#private'] ?? $sshParameters->getKeys()['private'];
+        $sshParameters->setPrivateKey($privateKey);
+        $sshParameters->setRemoteHost($dbConfig->getHost());
+        $sshParameters->setRemotePort($dbConfig->getPort());
+
+        if (!$sshParameters->getUser()) {
+            $sshParameters->setUser($dbConfig->getUser());
+        }
+
+        $tunnelParams = $sshParameters->toArray();
+        $this->getLogger()->info("Creating SSH tunnel to '" . $tunnelParams['sshHost'] . "'");
+        $proxy = new RetryProxy(
+            $this->getLogger(),
+            RetryProxy::DEFAULT_MAX_TRIES,
+            RetryProxy::DEFAULT_BACKOFF_INTERVAL,
+            ['SSHException', 'Exception']
+        );
+        try {
+            $proxy->call(function () use ($tunnelParams):void {
+                $ssh = new SSH();
+                $ssh->openTunnel($tunnelParams);
+            });
+        } catch (SSHException $e) {
+            throw new UserException($e->getMessage(), 0, $e);
+        }
+
+        $dbConfig->setHost('127.0.0.1');
+        $dbConfig->setPort($sshParameters->getLocalPort());
+        return $dbConfig;
+    }
+
+    protected function getConfigDefinitionClass(): string
+    {
+        $configRaw = $this->getRawConfig();
+        $action = $configRaw['action'];
+        $isConfigRow = !isset($configRaw['parameters']['tables']);
+
+        if ($action !== 'run') {
+            return ActionConfigDefinition::class;
+        } elseif ($isConfigRow) {
+            return ConfigRowDefinition::class;
+        } else {
+            return ConfigDefinition::class;
+        }
+    }
+
+    protected function getOutputFileName(string $outputTableName): string
+    {
+        $sanitizedTableName = Strings::webalize($outputTableName, '._');
+        return $sanitizedTableName . '.csv';
+    }
+
+    protected function getOutputFilePath(string $outputTableName): string
+    {
+        return $this->getDataDir() . '/out/tables/' . $this->getOutputFileName($outputTableName);
+    }
+
+    protected function checkConnectionIsAlive(): void
+    {
+        try {
+            $this->testConnection();
+        } catch (\Throwable $e) {
+            throw new DeadConnectionException("Dead connection: " . $e->getMessage());
+        }
+    }
+
+    protected function validateIncrementalFetching(
+        TableDetailParameters $table,
+        string $columnName,
+        ?int $limit = null
+    ): void {
+        throw new UserException('Incremental Fetching is not supported by this extractor.');
+    }
+
+    protected function quote(string $obj): string
```

Comment:
I wonder if we should maybe have 2 methods for this.  1) `quoteIdentifier` like this, and 2) `quoteValue` which we can use in queries to quote values.

Reply:
Yeah maybe renaming to `quoteIdentifiers` will be more descriptive. The `quoteValue` is provided via `PDO` for common use.